File: API/2281_Multik_SMA_Exp/CS/MultikSmaExpStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Contrarian strategy based on moving average slope.
/// Buys when SMA decreases for two consecutive segments and sells when it increases.
/// Positions are closed when the SMA slope reverses.
/// </summary>
public class MultikSmaExpStrategy : Strategy
{
	private readonly StrategyParam<int> _period;
	private readonly StrategyParam<DataType> _candleType;

	private decimal? _ma0;
	private decimal? _ma1;
	private decimal? _ma2;
	private decimal? _ma3;

	/// <summary>
	/// Moving average period.
	/// </summary>
	public int Period
	{
		get => _period.Value;
		set => _period.Value = value;
	}

	/// <summary>
	/// Candle type used for calculations.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Constructor.
	/// </summary>
	public MultikSmaExpStrategy()
	{
		_period = Param(nameof(Period), 50)
			.SetGreaterThanZero()
			.SetDisplay("MA Period", "Length of the moving average", "General")
			.SetCanOptimize(true)
			.SetOptimize(10, 100, 10);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to use", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_ma0 = _ma1 = _ma2 = _ma3 = null;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		var sma = new SMA { Length = Period };

		var subscription = SubscribeCandles(CandleType);

		subscription
			.Bind(sma, (candle, smaValue) =>
			{
				if (candle.State != CandleStates.Finished)
					return;

				if (!IsFormedAndOnlineAndAllowTrading())
					return;

				_ma3 = _ma2;
				_ma2 = _ma1;
				_ma1 = _ma0;
				_ma0 = smaValue;

				if (_ma3 is null || _ma2 is null || _ma1 is null)
					return;

				var dsma1 = _ma1.Value - _ma2.Value;
				var dsma2 = _ma2.Value - _ma3.Value;

				// Exit conditions
				if (dsma1 > 0 && Position > 0)
					SellMarket(Position);
				if (dsma1 < 0 && Position < 0)
					BuyMarket(-Position);

				// Entry conditions
				if (dsma2 < 0 && dsma1 < 0 && Position <= 0)
					BuyMarket(Volume + Math.Abs(Position));
				else if (dsma2 > 0 && dsma1 > 0 && Position >= 0)
					SellMarket(Volume + Math.Abs(Position));
			})
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, sma);
			DrawOwnTrades(area);
		}
	}
}

File: API/2281_Multik_SMA_Exp/PY/multik_sma_exp_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class multik_sma_exp_strategy(Strategy):
    def __init__(self):
        super(multik_sma_exp_strategy, self).__init__()
        self._period = self.Param("Period", 50) \
            .SetDisplay("MA Period", "Length of the moving average", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles to use", "General")
        self._ma0 = None
        self._ma1 = None
        self._ma2 = None

    @property
    def period(self):
        return self._period.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(multik_sma_exp_strategy, self).OnReseted()
        self._ma0 = None
        self._ma1 = None
        self._ma2 = None

    def OnStarted(self, time):
        super(multik_sma_exp_strategy, self).OnStarted(time)
        self._ma0 = None
        self._ma1 = None
        self._ma2 = None
        sma = ExponentialMovingAverage()
        sma.Length = self.period
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(sma, self.process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, sma)
            self.DrawOwnTrades(area)

    def process_candle(self, candle, sma_value):
        if candle.State != CandleStates.Finished:
            return
        sma_value = float(sma_value)
        self._ma2 = self._ma1
        self._ma1 = self._ma0
        self._ma0 = sma_value
        if self._ma2 is None or self._ma1 is None:
            return
        dsma1 = self._ma0 - self._ma1
        dsma2 = self._ma1 - self._ma2
        if dsma2 < 0 and dsma1 < 0 and self.Position <= 0:
            self.BuyMarket()
        elif dsma2 > 0 and dsma1 > 0 and self.Position >= 0:
            self.SellMarket()

    def CreateClone(self):
        return multik_sma_exp_strategy()
